docs/plugins.md

```diff
@@ -0,0 +1,79 @@
+# Developing 3scale Toolbox CLI plugins
+
+3scale Toolbox CLI is based on [cri](https://github.com/ddfreyne/cri) library for building command line tools.
+Plugin system also uses [cri](https://github.com/ddfreyne/cri) to leverage easy to develop, flexible and extensible plugin system.
+
+3scale Toolbox will load plugins installed in gems or $LOAD_PATH. Plugins are discovered via *Gem::find_files*, then loaded.
+Few requirements must be met for a plugin to be loaded:
+* Plugin must be named `${plugin_name}_command.rb` and placed at `/3scale_toolbox/commands/` from your root load path
+* Plugin must implement module function `ThreeScaleToolbox::Commands.command_${plugin_name}_definition` and return instance of `Cri::Command` from [cri](https://github.com/ddfreyne/cri)
+
+Nothing better than `simple Hello World plugin` to illustrate.
+
+```
+$ cat lib/3scale_toolbox/commands/foo_command.rb
+module ThreeScaleToolbox
+  module Commands
+    def self.command_foo_definition
```

Comment:
I don't think it is a good idea to let plugins reopen this module. 

I think a class/module should be defined in one file and thats it. Then in theory it could be frozen, so no other files can reopen it. Not that we would actually do so, but it sets some expectations and mindset.

I think more reasonable API would be to let submodules register into some command set. 

For example if you require `foo_command` then it could at the and of the file/module add itself to 3scale toolbox `ThreeScaleToolbox::CLI.add_subcommand self`. 

This would allow extending (copy service, copy account, copy provider, ...) non root (copy subcommand) commands like said in https://github.com/3scale/3scale_toolbox/pull/50/files#r201001411.

I think we should have an API that works on any level, not just the first one.
And one that uses own namespace and public API instead of reopening modules from different gem.

Reply:
Both issues you underline were intentionally defined as they are based on RubyGems plugin system design. 
* Only extending first level commands 
* Placing commands under a known namespace [here](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/command.rb#L606)
```
##                                            
# \Commands will be placed in this namespace  
                                              
module Gem::Commands                          
end
```
In fact, all rubygems built-in commands are defined in one file each and all of them under ```Gem::Commands``` namespace.

WRT only extending at command level (not allowing extending subcommands), it is indeed less flexible. But Rubygems does not allow it (not sure if heroku allows it either) and seemed to me a good starting point. Besides, it felt to me that subcommands should belong (at least in 99% of use cases) to main comand plugin project. 

If extending subcommands is a requirement, we should highlight it clearly. It was not in requirements initially.

Can you elaborate more on your  idea about letting submodules register into some command set? I'll try to figure out how to do it.

Reply:
I guess you mean

```
$ cat lib/3scale_toolbox/commands/foo_command.rb
module MyModuleName
  def self.command_definition
    Cri::Command.define do
      name        'foo'
      usage       'foo [options]'
      summary     'foo command'
      description 'This command does a lot of stuff.'
      run do |opts, args, cmd|
        puts "Hello World"
      end
    end
  end
end

ThreeScaleToolbox::CLI.add_command MyModuleName.command_definition

$ RUBYOPT=-Ilib 3scale foo
Hello World
```

This is a little change. No problem to do it. Although I like the way it is now. In order to discover plugins automatically, there needs to be some well known filepath pattern that is passed to ```Gem::find_files```. Right now it is defined as: ```3scale_toolbox/commands/*_command.rb```, so using modules like ```MyModuleName``` in a file called ```3scale_toolbox/commands/foo_command.rb``` does not seem right.

But still we need to define a way to extend on any level.

Reply:
@eguzki Yes, I meant that. I'm not against letting plugins use the 3scale_toolbox namespace. That could be easily lifted in the future. My main beef is reopening a module to add there a method that does not belong to the namespace of the library. In the future we can look for file named `3scale_toolbox_plugin.rb` in any of the gems and just load it.

With this pattern you should get the subcommands almost for free. They can register themselves into the main command or the main command can register them. 

```ruby
module ThreeScaleToolbox::Commands::Copy
  require_relative './copy/service'
  require_relative './copy/account'
end
```

```ruby
module ThreeScaleToolbox::Commands::Copy::Service
  ...
  ThreeScaleToolbox::Commands::Copy.add_command self
end
```

```ruby
module ThreeScaleToolbox::Commands::Copy::Account
  ...
  ThreeScaleToolbox::Commands::Copy.add_command self
end
```

Of course it is less than ideal because you actually have to require them, but that is not a big deal when you are in a scope of one module/command. Plugins packaged as gems could easily register themselves anywhere.